### PR TITLE
[drape] Make lines great again! :)

### LINKED ISSUE #5414: Dashed lines are semi-transparent?
Dashed lines seem to have some transparency?
Although in the style they're set to be `opacity: 1`:

2023-06-29-235647_1366x768_scrot

The most "transparent" or light line has only one type `highway-path`.
The most dark lines have 3, e.g. `highway-path-bicycle highway-path-horse highway-path-hiking` - this makes the same line (because these types share the very same style) to be rendered three times! Hence the color becomes more dark (as its intended be, actually).
I reproduced this bug on qt/desktop and on android emulator.

Could it be because of the way dashed lines are rendered using masked textures? @vng 

Update: I'm not sure its happening on my android phone though - if there is a difference, then its very slight. So maybe it depends on visual scale (my desktop and emulator are 1.0 and 1.5 respectively). So maybe its because of AA-like rendering of lines - their borders become semi-transparent and on low dpi screens its more bordering semi-transparent pixels than the core ones..
Btw I tested switching the AA on/off in the desktop app - it doesn't have an effect on this bug. Then it might be related to #4467.

### LINKED ISSUE #6583: Fix blurry thick lines
Thicker and/or dashed lines look blurry/dirty compared to thin solid lines, e.g.
(magnified several times to show the issue)

image

Thin black line is very crisp, while only the core of the dashed line has a true color and the outer parts are semi-transparent.
Causes
- #5414.

For thinner solid lines OM uses GL line primitives, while dashed or thicker lines are drawn by explicitly-set textured quads
(see `drape_frontend/line_shape.cpp LineShape::Prepare()`).

Maybe this blurriness was added intentionally as a kind of anti-aliasing? But it looks bad, especially on lower dpi screens.

CC @vng

### LINKED ISSUE #6901: Isolines are not visible in the dark theme
Especially Outdoors style should highlight them.

### LINKED ISSUE #6937: Contour lines are too prominent in non-outdoors style
The General style can be more readable if contour lines will have similar colors as in outdoors style.

General | Outdoors
--------|---------
выява|выява

### PR DESCRIPTION
**Pull request by pastk**
This PR disabled the ad-hoc AA that was implemented via shaders:
- gives full lines' contrast / color saturation control to style designers
- makes contrast / color saturation consistent and predictable (exactly as its set in the style) for various lines drawing methods (custom or via driver's primitives), widths and screen densities; and consistent with triangles rendering
- more crisp and clean looks (atm lines edges are blurred)
- some (probably minor) performance improvements / battery saving

Lines' opacities were adjusted to retain the original "soft" look of OM styles,
with some exceptions/improvements:
- returned a slightly greyish color we had for smaller roads and footways before #2134; it makes secondaries a bit more pronounced compared to tertiaries and lower
- a bit more contrast on the World map
- vehicle: made railways and tramlines slightly less pronounced than before
- outdoors (we lacked contrast in many cases before):
  * more contrast to all roads and railways
  * paths and tracks
    * on z11 no contrast increase as there are cases when there are too many lines
    * on z12 a minor increase
    * full power from z13-
  * minor cutlines and drains/ditches improvements
- increased isolines' contrast for the outdoors and reduced for the main style
- balanced isolines' contrast in dark modes

Check screenshots starting from https://github.com/organicmaps/organicmaps/pull/6590#issuecomment-1817918391

- fixes #6583
- fixes #5414
- fixes #6901
- fixes #6937
- improves #2123
- related to #4817

<details>
<summary>Original first post:</summary>

OM renders thicker and/or dashed lines by creating many quads/triangles, while using GL line primitives fallback for thinner solid lines only.
The latter is rendered very crisp, while there is a very aggressive AA-like blurring of edges for other lines.

A magnified sample from a low dpi desktop screen:
img

A magnified sample from a high dpi android device:
img

After disabling this "anti-aliasing":
image

Basically it makes many lines semi-transparent, pale, and blurry / not clean looking.

z19 high dpi
(animated gifs)
z19-highdpi

z14 high dpi
z14-highdpi

z15 low dpi
z15-desktop

Note how thin roads and cycleways (closer to the port) didn't change - they're rendered by GL line primitives apparently.

Some lines become too prominent maybe with this change. Its possible to correct their color, opacity, thickness. But maybe its just a matter of getting used to it.
</details>

### CODE REVIEW

**Comment by vng** [comment]
AA lines look better, IMO.

**Comment by pastk** [comment]
There is an option to adjust blur params to make it much less aggressive, but I think its better to disable it completely:
- to make it consistent with other (thin solid) lines and areas
- the visual quality improvement even for low dpi screens is not significant, on high dpi devices its barely noticeable
- less work for fragment shaders, better battery life

**Comment by pastk** [comment]
> AA lines look better, IMO.

Why? Maybe they look too contrast now because we all used to the current "soft" style? Its better adjusted by styling.

When this kind of AA is used there is no way to unlock max lines contrast needed in many situations, so we used to workaround by making lines thicker or colors brighter.

It leads to bugs also, e.g. https://github.com/organicmaps/organicmaps/issues/5414

**Comment by vng** [comment]
Yes, maybe because of softer colors. And yes, probably we can tune styles and the picture will be better for me.
Is it fast to fix and compare for path, footway, cycleway, service?

**Comment by pastk** [comment]
A quick test with "softened" paths, tracks, footways, ferry lines (I've left the roads full power white).

I set them `opacity: 0.5;` (sic!) and they're still a tad more contrast than the master!
It shows how much contrast and color space this AA kept locked from us.

z15-desktop-opacity05

Note that e.g. a small leisure-track near the stadium is also as soft and consistent as other lines around (its 0.5 opaque also) - in master its way too pronounced and inconsistent compared to other lines (as its drawn by gl primitives).

**Comment by pastk** [comment]
Another serious issue with this AA approach is that its effect depends on line thickness a lot - as it blurs "the edge" only, but "edge" width is set very big, so all pixels of thin lines get semi-transparent.

Hence thinner lines are affected the most and they loose a lot of contrast / color saturation.
While thick (in pixels) lines retain their core unaffected, hence more contrast and color.

This also means there is inconsistency in lines representation for different density devices. The AA affects high dpi devices less (as same lines have bigger pixel width there), so users of low dpi devices face more faded and less contrast lines.

E.g. in previous post I used opacity=0.5 to match AA's "soft" effect for z15 on vs=1 device, while for vs=3 device opacity=0.6 was enough to achieve the same effect.

**Comment by vng** [comment]
Ok, I agree that hardcoded AA is not good. Can at least be set for specific types with a special style's flag.

**Comment by pastk** [comment]
Just for comparison.
Master vs "proper" driver-side `?aa` (our "ad-hoc AA" is disabled, no opacity adjustments).
As one can see the "proper" AA smoothens "jaggies" yet doesn't lead to contrast or color loss.
This is on vs=1 low dpi screen.

z15-desktop-systemAA

**Comment by biodranik** [comment]
With this patch, outdoor style is not needed amymore. The base style becomes an outdoor one. I don't like how the map looks now, way too much contrast from tracks on Z14.

выява

Note how bad are white dashed lines
выява

**Comment by pastk** [comment]
> With this patch, outdoor style is not needed amymore.

This patch doesn't make trails visible on higher zooms 😂
But you're right that the outdoors style will especially benefit from it (w/o increasing lines widths). Maybe we won't even need a "high-contrast" option then.

> The base style becomes an outdoor one. I don't like how the map looks now, way too much contrast from tracks on Z14.

Did you see the "softened" version?
https://github.com/organicmaps/organicmaps/pull/6590#issuecomment-1816579486

Getting rid of this ad-hoc AA will give style designers full control of contrast and color saturation, we'll be able to set it low or high depending on the style or particular line feature type. Atm we're locked to the "low" option...

**Comment by pastk** [comment]
BTW what is the visual scale of your desktop screen?

**Comment by biodranik** [comment]
> BTW what is the visual scale of your desktop screen?

Visual scale = 2 ; Tile size = 1024 ; Resources = xhdpi

> Getting rid of this ad-hoc AA will give style designers full control of contrast and color saturation, we'll be able to set it low or high depending on the style or particular line feature type. Atm we're locked to the "low" option...

I agree. With an additional benefit of saving the battery without AA.
So the right question is: how can we implement (almost) the same low-contrast generic style without AA now?

**Comment by pastk** [comment]
A summary of pros and cons I see.

Pros:
- gives full lines' contrast / color saturation control to style designers
- makes contrast / color saturation consistent and predictable (exactly as its set in the style) for various lines drawing methods (custom or via driver's primitives), widths and screen densities; and consistent with triangles rendering
- more crisp and clean looks (atm lines edges are blurred)
- some (probably minor) performance improvements / battery saving

Cons:
- some more visible "jaggies" on low dpi devices (they were always present for all areas edges and some lines too)
- needs re-styling work to avoid too contrast and promiment lines to retain the original "softer" looks, especially at low country zooms

For now a reasonable way would be to reduce lines' opacities via styles to match similar amount of transparency the ad-hoc AA currently produces. 
But IMHO in the long term generally its better to tune lines' colors and/or thicknesses and use transparency when its really needed only.

**Comment by pastk** [comment]
> So the right question is: how can we implement (almost) the same low-contrast generic style without AA now?

With some styles-added transparency, see above.

Maybe a much smaller amount of AA could be useful still too emulate finer line thickness controls, as otherwise its quite coarse and limited to (sub)pixel steps. I'll do some experiments to see if this AA actually will help here...

**Comment by biodranik** [comment]
> But IMHO in the long term generally its better to tune lines' colors and/or thicknesses and use transparency when its really needed only.

The current base OM style is great, many users get used to it. It really needs to stay as polished as it is now.

Btw, did you check commits/PR history why AA was introduced in the first place?

**Comment by pastk** [comment]
> Btw, did you check commits/PR history why AA was introduced in the first place?

`[drape] manual line anti-aliasing` but no info why it was made so aggressive (2.5 pixels)
https://github.com/mapsme/omim/commit/b4bbe6484b44925f3572e298769c69156fc10309

> Maybe a much smaller amount of AA could be useful still too emulate finer line thickness controls, as otherwise its quite coarse and limited to (sub)pixel steps. I'll do some experiments to see if this AA actually will help here...

It doesn't really work as intended.
Visual thickness do change with fractional line width increases, but just enabling this AA reduces visual thickness / contrast of the line noticeably (even when AA is set to 0.5 pixels, and this setting makes all [0-1] width lines look the same, so its start working from >1 widths only).
I'm pretty sure its possible to change the algo to take "virtual" width into account properly, but it requires better understanding of shaders.

**Comment by pastk** [comment]
A good example of related rendering inconsistency bug.

Master build, same android device, GL vs Vulkan rendering:

tramline-bug

Vulkan: tram lines are very faded ("AA" makes them very transparent)
GL: tram lines are very visible as they're drawn by GL primitives and hence unaffected by "A" (the driver reports Max line width = 8, while in Vulkan its = 1 so can't be used for this line)

Same likely happens with other thin solid lines like cycleways.

**Comment by pastk** [comment]
Comparison of master vs disabled ad-hoc AA compensated by opacity and/or widths adjustments (on a vs=3 device).

In animated gif format. If you hardly see a difference it means I did a good job :)

I returned a slightly greyish color we had for smaller roads and footways before #2134. It makes secondaries a bit more pronounced compared to tertiaries and lower.
noaa-z11

noaa-z15

noaa-z18

IMHO its good to have a bit more contrast on the World map, WDYT?
noaa-z8

**Comment by biodranik** [comment]
Looks good. @pastk it would be great to test your changes in the PR.

**Comment by pastk** [comment]
- regenerated Vulkan shaders, so AA is disabled there too
- disabled AA in Metal shaders - please test on iOS!
- the main style is good to test (see above screenshots)

TODO: tune vehicle and outdoors styles; check dark themes

**Comment by pastk** [comment]
Note: need to run `tools/unix/generate_vulkan_shaders.sh` to regenerate vulkan shaders. Would be good to trigger it automatically on shaders' sources change.

**Comment by pastk** [comment]
It also "fixes" the bug with water being visible around road edges (more noticeable on lower dpi)
image

After the change:
image

But the actual root of this bug is
- #5339

**Comment by pastk** [comment]
"Softened" the vehicle style too.
- made railways and tramlines slightly less pronounced than before
- added ferry lines, but less pronounced than in main style
- fixed bug that made e.g. `highway-path psurface-unpaved_bad` appear at z14- as white dash while ordinary `highway-path`s are z17-

noaa-veh-z11

noaa-veh-z14

noaa-veh-z18

**Comment by pastk** [comment]
Processed the outdoors style too.
- more contrast to all roads and railways
- paths and tracks
  - on z11 no contrast increase as there are cases when its too much already
  - on z12 a minor increase
  - full power from z13-
- increased visibility of fences, ditches and bridleways by one zoom compared to the main style

noaa-od-z12

noaa-od-z13

noaa-od-z15

noaa-od-z17

**Comment by pastk** [comment]
Minor cutlines and drains/ditches improvements:
noaa-od-z13c

Checked that dark/night mode didn't break too, here is vehicle night style:
noaa-veh-z15d

**Comment by pastk** [comment]
Please test!

**Comment by dvdmrtnz** [comment]
I have tested these changes for a bit. They pretty much look the same to me. The only thing that I noticed is that the white parking lanes seem to blend more with the yellow parking background.

Before / After:

**Comment by pastk** [comment]
Thanks for testing!

> The only thing that I noticed is that the white parking lanes seem to blend more with the yellow parking background.

Right. This is because of 

> I returned a slightly greyish color we had for smaller roads and footways before https://github.com/organicmaps/organicmaps/pull/2134. It makes secondaries a bit more pronounced compared to tertiaries and lower.

You've tested on iPhone, right? Looks like shaders work as expected there, which is great!
Did you notice higher contrast on e.g. paths in the outdoors style? Should be mostly noticeable around z13 - z16

**Comment by biodranik** [comment]
Does it improve the contrast in the dark mode?

**Comment by pastk** [comment]
> Does it improve the contrast in the dark mode?

In the current state the changes are minimal as by design, see a screenshot https://github.com/organicmaps/organicmaps/pull/6590#issuecomment-1824213963

**Comment by pastk** [comment]
In this PR isolines were made too prominent, especially visible in the dark mode.
Also on master isolines in the main style were more prominent than in the Outdoors.

Tuned.
Now compared to master:
- main style isolines somewhat less prominent
- outdoors style isolines somewhat more prominent (but not too much)
- outdoors isolines are slightly more prominent that main styles ones

Before/after comparison with master.
Main, dark:
iso-dark

Outdoors, dark:
iso-od-dark

Outdoors, light:
iso-od-light

**Comment by biodranik** [comment]
Can the isolines patch be merged in a separate PR?

**Comment by pastk** [comment]
> Can the isolines patch be merged in a separate PR? 

No, it works together with extra contrast capabilities this PR provides.

**Comment by pastk** [comment]
I've removed some not directly related visibility changes (bridleways, fences, unpaved types fix...) and updated the first post with a summary.

PTAL!
@vng @biodranik

**Review by biodranik — APPROVED** [review]
Thanks! To me, it now looks similar to the general style, except tracks, that are a bit more prominent (although acceptable).

LGTM, @vng WDYT?

Is it possible to measure performance/battery usage on some dense areas in outdoors style with and without AA?

**Comment by pastk** [comment]
> 
> Is it possible to measure performance/battery usage on some dense areas in outdoors style with and without AA?

I doubt the difference will be measurable, especially with our current bench-marking tools. Maybe with some special tool that shows GPU utilization?

**Review by vng — APPROVED** [review]
